g2p/api_v2.py

```diff
@@ -0,0 +1,406 @@
+
+
+@api.post("/convert")
+def convert(  # noqa: C901
+    request: ConvertRequest = Body(
+        examples={
+            "eng-ipa to eng-arpabet": {
+                "summary": "Convert English IPA to ARPABET",
+                "description": "G2P can do simple conversions between equivalent phonetic notations",
+                "value": {
+                    "in_lang": "eng-ipa",
+                    "out_lang": "eng-arpabet",
+                    "text": "hɛloʊ",
+                },
+            },
+            "fin to eng-arpabet": {
+                "summary": "Convert Finnish orthography to nearest ARPABET phones",
+                "description": """
+G2P can do also (sometimes) approximate one language's phonology using phones from another.
+To find out possible output languages for an input, use the 'outputs_for' endpoint.""",
+                "value": {
+                    "in_lang": "fin",
+                    "out_lang": "eng-arpabet",
+                    "text": "hyvää yötä",
+                },
+            },
+            "composed conversion": {
+                "summary": "Convert Finnish orthography directly to nearest ARPABET phones",
+                "description": "By default all the conversion steps are returned.  "
+                "You can get the direct mapping from input to output by setting compose_from: in_lang",
+                "value": {
+                    "in_lang": "fin",
+                    "out_lang": "eng-arpabet",
+                    "text": "hyvää huomenta",
+                    "compose_from": "fin",
+                },
+            },
+            "tokenized conversion": {
+                "summary": "Convert Finnish orthography with punctuation directly to nearest ARPABET phones",
+                "description": "Non-word segments are returned with in_lang and out_lang as null",
+                "value": {
+                    "in_lang": "fin",
+                    "out_lang": "eng-arpabet",
+                    "text": "Varaus! Sieni, joka kasvaa korvessa, on myrkyllinen!",
+                    "compose_from": "fin-ipa",
+                },
+            },
+        }
+    )
+) -> List[Segment]:
+    """Tokenize a text return the converted and intermediate forms of each
+    segment (non-token segments will have converted=False).  The final
+    conversion comes first in the output, followed by prevoius
+    conversions.  If you do not want the intermediate conversions, set
+    "compose_from" to in_lang.
+
+    """
+    in_lang = request.in_lang.name
+    out_lang = request.out_lang.name
+    try:
+        transducer = g2p.make_g2p(in_lang, out_lang)
+        if request.tokenize:
+            tokenizer = g2p.make_tokenizer(in_lang)
+            tokens = tokenizer.tokenize_text(request.text)
+        else:
+            tokens = [{"text": request.text, "is_word": True}]
+    except NoPath:
+        raise HTTPException(
+            status_code=400, detail=f"No path from {in_lang} to {out_lang}"
```

Comment:
Pylint suggests doing `except NoPath as exc` and `raise ... from exc`. I still don't fully understand what that changes, but I tend to just do it when it reminds me.

Reply:
Ah, wait, reading up on this, the better solution here would be `raise HTTPException() from None` because the caller does not need to know that the cause is our internal `NoPath` exception, they only need to get our 400`.
Informative reading: https://stackoverflow.com/questions/24752395/python-raise-from-usage